Comments and undef

cxw42 opened the pull request:
This PR builds on #2, and fixes #3.  It also includes a basic fix for #4.  Again, I have not updated `Changes` or the version number, but am happy to do so.

- #3: `<!--...-->` strings will now create comments instead of text nodes.
- #4: `undef` values for node text or children will now be ignored.

Thank you for considering this PR!

salva reviewed (commented):
I don't like having a magic marker as `<!-- -->` for text strings. After all, it is perfectly valid content.

I would prefer having the magic marker as a tag name. For instance, something like this:

    $xml = [ foo => {},
              ...
              ['!--', "This is a comment"] ]

Regarding your proposal to ignoring the undefs, I think that even if your use case is valid, other people would like to have something stricter. A good compromise would be to have the module generate a warning when some undef is found and let the user disable the warnings with a `no warnings 'XML::FromPerl';` pragma.

cxw42 commented:
@salva Thanks for reviewing!  Modified per your comments; coverage still 100%.

- `!--` for comment tags is implemented.  I found out `XML::LibXML::Comment->appendChild()` is a no-op, so I just skip any child nodes in a comment (and have documented that fact).
- I added the warnings on `undef`.
- I changed the logic to create a sentinel node ("phony") and append to that while recursing, rather than returning new nodes while recursing.  That way calling code doesn't have to check the return value from the recursive function (now `_fill_node_children()`).
- I also updated the Changes file but left the date blank.

Let me know what you think!  Thanks!

cxw42 committed:
Create comments; skip undefs

- Create comment nodes with the tag name `!--` as requested by salva.
  Fixes #3.  Skips undef children and nested nodes, and joins text
  using `$,`.
- Skip `undef` children.  Fixes #4.
- As requested by salva, create a new warnings category
  `XML::FromPerl::undefined`.  Warns on undef if categories
  XML::FromPerl and XML::FromPerl::undefined are enabled.

This commit reworks the recursion.  Instead of building a new node and
returning it, each call to new function _fill_node_children() adds zero
or more children to an already-existing node.  That way you do not have
to special-case _not_ adding a child of a node.

cxw42 commented:
I need to make some changes - <5.14 doesn't support custom warning categories, so I need to tweak accordingly.  [perldoc](https://perldoc.perl.org/5.12.0/warnings.html), [Travis](https://travis-ci.org/cxw42/p5-XML-FromPerl/builds/544706684)

cxw42 committed:
Warn on undef; custom warning categories

cxw42 commented:
OK - I have made changes and all tests now pass!  https://travis-ci.org/cxw42/p5-XML-FromPerl/jobs/549119725 .  What say you?  Thanks!

cxw42 commented:
@salva Any thoughts?  Thanks!

salva commented:
@cxw42,  sorry for the delay, nowadays I can't devote too much time to the maintenance of my Perl modules.

I have tried to review your changes several times, but the thing that really stops me for approving them is that you have included three different things in the same commit, and so, it is difficult to me to review them.

Specifically I am not completely convinced for the convenience of using a dummy node while creating the tree.

It would be much easier if you divide the changes in three different commits.

Regards
